File: src/main/java/com/gmail/borlandlp/minigamesdtools/lobby/LobbyHubController.java
package com.gmail.borlandlp.minigamesdtools.lobby;

import com.gmail.borlandlp.minigamesdtools.APIComponent;
import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools;
import com.gmail.borlandlp.minigamesdtools.config.ConfigEntity;
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath;
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider;
import org.bukkit.Bukkit;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.entity.Player;
import org.bukkit.event.HandlerList;
import org.bukkit.event.Listener;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.ArrayList;
import java.util.List;

public class LobbyHubController implements APIComponent, LobbyServerAPI {
    private List<ServerLobby> lobbies = new ArrayList<>();
    private BukkitTask task;
    private Listener lobbyListener;
    private ServerLobby defaultServerLobby;
    private boolean startLobbyEnabled;

    public boolean isStartLobbyEnabled() {
        return startLobbyEnabled;
    }

    public ServerLobby getLobbyByPlayer(Player p) {
        for (ServerLobby serverLobby : this.lobbies) {
            if(serverLobby.contains(p)) return serverLobby;
        }

        return null;
    }

    public ServerLobby getLobbyByID(String id) {
        for (ServerLobby serverLobby : this.lobbies) {
            if(serverLobby.getID().equals(id)) return serverLobby;
        }

        return null;
    }

    public void register(ServerLobby l) {
        this.lobbies.add(l);
        l.onRegister();
    }

    public void unregister(ServerLobby l) {
        this.lobbies.remove(l);
        try {
            l.onUnregister();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public ServerLobby getDefaultServerLobby() {
        return defaultServerLobby;
    }

    private void doWork() {
        for (ServerLobby serverLobby : this.lobbies) {
            try {
                serverLobby.tickUpdate();
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
    }

    public void onLoad() {
        final LobbyHubController task = this;
        this.task = new BukkitRunnable() {
            public void run() {
                task.doWork();
            }
        }.runTaskTimer(MinigamesDTools.getInstance(), 0, 20);

        this.lobbyListener = new LobbyListener(this);
        Bukkit.getServer().getPluginManager().registerEvents(this.lobbyListener, MinigamesDTools.getInstance());

        //load all serverLobby
        for (ConfigEntity configEntity : MinigamesDTools.getInstance().getConfigProvider().getPoolContents(ConfigPath.SERVER_LOBBY)) {
            if(configEntity.getData().get("enabled").toString().equals("true"))
            try {
                Debug.print(Debug.LEVEL.NOTICE, "Load ServerLobby[ID:" + configEntity.getID() + "]");
                this.register(MinigamesDTools.getInstance().getLobbyCreatorHub().createLobby(configEntity.getID(), new DataProvider()));
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        // set default serverLobby
        ConfigurationSection conf = null;
        try {
            conf = MinigamesDTools.getInstance().getConfigProvider().getEntity(ConfigPath.MAIN, "minigamesdtools").getData();
        } catch (Exception e) {
            e.printStackTrace();
        }
        // а есть ли смысл в валидации на нуллпоинетер для conf? всё равно, этот случай - критичен и должен быть отловлен вышестоящим загрузчиком.

        this.startLobbyEnabled = Boolean.parseBoolean(conf.get("start_lobby.move_on_server_join").toString());
        ServerLobby serverLobby = this.getLobbyByID(conf.get("start_lobby.id").toString());
        if(serverLobby != null) {
            this.defaultServerLobby = serverLobby;
        } else {
            Debug.print(Debug.LEVEL.NOTICE, "Cant find default server lobby!");
            return;
        }

        // fix on 'reload' command
        for (Player player : Bukkit.getServer().getOnlinePlayers()) {
            if(this.getLobbyByPlayer(player) == null) {
                this.getDefaultServerLobby().registerPlayer(player);
            }
        }
    }

    public void onUnload() {
        this.task.cancel();

        HandlerList.unregisterAll(this.lobbyListener);

        (new ArrayList<>(this.lobbies)).forEach(this::unregister);
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/lobby/LobbyHubController.kt
package com.gmail.borlandlp.minigamesdtools.lobby

import com.gmail.borlandlp.minigamesdtools.APIComponent
import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.DefaultCreators
import com.gmail.borlandlp.minigamesdtools.MinigamesDTools
import com.gmail.borlandlp.minigamesdtools.config.ConfigPath
import com.gmail.borlandlp.minigamesdtools.creator.DataProvider
import org.bukkit.Bukkit
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.entity.Player
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import org.bukkit.scheduler.BukkitRunnable
import org.bukkit.scheduler.BukkitTask
import java.util.function.Consumer
import kotlin.collections.ArrayList

class LobbyHubController : APIComponent, LobbyServerAPI {
    private val lobbies: MutableList<ServerLobby> = ArrayList()
    private var task: BukkitTask? = null
    private var lobbyListener: Listener? = null
    var defaultServerLobby: ServerLobby? = null
        private set
    var isStartLobbyEnabled = false
        private set

    override fun getLobbyByPlayer(player: Player): ServerLobby? {
        for (serverLobby in lobbies) {
            if (serverLobby.contains(player)) return serverLobby
        }
        return null
    }

    override fun getLobbyByID(id: String): ServerLobby? {
        for (serverLobby in lobbies) {
            if (serverLobby.id == id) return serverLobby
        }
        return null
    }

    override fun register(serverLobby: ServerLobby) {
        lobbies.add(serverLobby)
        serverLobby.onRegister()
    }

    override fun unregister(serverLobby: ServerLobby) {
        lobbies.remove(serverLobby)
        try {
            serverLobby.onUnregister()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun doWork() {
        for (serverLobby in lobbies) {
            try {
                serverLobby.tickUpdate()
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }
    }

    override fun onLoad() {
        val task = this
        this.task = object : BukkitRunnable() {
            override fun run() {
                task.doWork()
            }
        }.runTaskTimer(MinigamesDTools.instance, 0, 20)
        lobbyListener = LobbyHubListener(this)
        Bukkit.getServer().pluginManager.registerEvents(lobbyListener, MinigamesDTools.instance)
        //load all serverLobby
        for (configEntity in MinigamesDTools.instance!!.configProvider!!.getPoolContents(ConfigPath.SERVER_LOBBY)) {
            if (configEntity.data["enabled"].toString() == "true") try {
                Debug.print(
                    Debug.LEVEL.NOTICE,
                    "Load ServerLobby[ID:" + configEntity.id + "]"
                )
                register(
                    MinigamesDTools.instance!!.creatorsRegistry.get(DefaultCreators.SERVER_LOBBY.pseudoName)!!.create(
                        configEntity.id,
                        DataProvider()
                    ) as ServerLobby
                )
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        // set default serverLobby
        var conf: ConfigurationSection? = null
        try {
            conf = MinigamesDTools.instance!!.configProvider!!.getEntity(ConfigPath.MAIN, "minigamesdtools")!!
                .data
        } catch (e: Exception) {
            e.printStackTrace()
        }
        // а есть ли смысл в валидации на нуллпоинетер для conf? всё равно, этот случай - критичен и должен быть отловлен вышестоящим загрузчиком.
        isStartLobbyEnabled = conf!!["start_lobby.move_on_server_join"] as Boolean
        val serverLobby = getLobbyByID(conf["start_lobby.id"].toString())
        if (serverLobby != null) {
            defaultServerLobby = serverLobby
        } else {
            Debug.print(
                Debug.LEVEL.NOTICE,
                "Cant find default server lobby!"
            )
            return
        }
        // fix on 'reload' command
        for (player in Bukkit.getServer().onlinePlayers) {
            if (getLobbyByPlayer(player) == null) {
                defaultServerLobby!!.registerPlayer(player)
            }
        }
    }

    override fun onUnload() {
        task!!.cancel()
        HandlerList.unregisterAll(lobbyListener)
        lobbies.toMutableList().forEach(Consumer(this::unregister))
    }
}